src/mkfs.fat.c

```diff
@@ -1028,6 +1029,100 @@ static void setup_tables(void)
 	}
     }
 
+    if (fill_mbr_partition) {
+        uint8_t *partition;
+        uint8_t *disk_sig_ptr;
+        uint32_t disk_sig;
+        uint8_t buf[512];
+        int fd;
+
+        if (verbose)
+            printf("Adding MBR table\n");
+
+        if (size_fat == 32)
+            disk_sig_ptr = bs.fat32.boot_code + BOOTCODE_FAT32_SIZE - 16*4 - 6;
+        else
+            disk_sig_ptr = bs.oldfat.boot_code + BOOTCODE_SIZE - 16*4 - 6;
+
+        if (*(disk_sig_ptr-1)) {
+            printf("Warning: message too long; truncated\n");
+            *(disk_sig_ptr-1) = 0;
+        }
+
+        disk_sig = 0;
+        memset(disk_sig_ptr, 0, 16*4 + 6);
+
+        /* Try to read existing 32 bit disk signature */
+        fd = open(device_name, O_RDONLY);
+        if (fd >= 0) {
+            if (read(fd, buf, sizeof(buf)) == sizeof(buf) && buf[510] == 0x55 && buf[511] == 0xAA)
+                disk_sig = (uint32_t)buf[440] | ((uint32_t)buf[441] << 8) | ((uint32_t)buf[442] << 16) | ((uint32_t)buf[443] << 24);
+            close(fd);
+        }
+
+        /* If is not available then generate random 32 bit disk signature */
+        if (invariant)
+            disk_sig = volume_id;
+        else if (!disk_sig)
+            disk_sig = generate_volume_id();
+
+        disk_sig_ptr[0] = (disk_sig >>  0) & 0xFF;
+        disk_sig_ptr[1] = (disk_sig >>  8) & 0xFF;
+        disk_sig_ptr[2] = (disk_sig >> 16) & 0xFF;
+        disk_sig_ptr[3] = (disk_sig >> 24) & 0xFF;
+
+        partition = disk_sig_ptr + 6;
+
+        /* Active flag */
+        partition[0] = 0x80;
+
+        /* CHS address of the first sector */
+        partition[1] = 0;
+        partition[2] = 0;
```

Comment:
```suggestion
        partition[2] = 1;
```
CHS sector numbers are 1-based.

Reply:
Good catch, thanks!

End sector was already correctly calculated.